### DISCUSSION TITLE
Option to prevent screen locking

### DISCUSSION THREAD

**Discussion by manio** [FreeRDP/FreeRDP#9473] [Show and tell]
Hi,
Sometimes it is desirable to prevent screen locking on the remote windows session. There are some ways to do so. The most obvious is probable to set it on the windows side ([like this](https://serverfault.com/questions/1034302/how-do-i-prevent-rdp-sessions-from-idle-locking)).

Unfortunately reconfiguring target windows system may be not possible (or not possible directly) and I found in such cases people are trying other methods like using xdotool/[ydotool](https://github.com/ReimuNotMoe/ydotool) automations or even starting VLC, configuring it to disable screensaver and starting a clip in a loop (sic!) ([like this](https://forum.devolutions.net/topics/1151/rdp-keep-alive#177358)) or enabling presentation mode.

Instead of trying such approach I prepared a different solution based on the RDP protocol itself. I am just sending a `input_send_mouse_event` with hardcoded cursor position (in the below case it is position x=800, y=600) in a constant time periods (in my case every 3 minutes).

Here are the changes I've made (currently it is on top of 2.11.2):
```diff
diff --git a/client/Wayland/wlfreerdp.c b/client/Wayland/wlfreerdp.c
index 65e29bc51..7576b7196 100644
--- a/client/Wayland/wlfreerdp.c
+++ b/client/Wayland/wlfreerdp.c
@@ -29,6 +29,7 @@
 #include <freerdp/gdi/gdi.h>
 #include <freerdp/client.h>
 #include <freerdp/utils/signal.h>
+#include <freerdp/utils/stopwatch.h>
 #include <freerdp/locale/keyboard.h>
 
 #include <linux/input.h>
@@ -472,6 +473,7 @@ static int wlfreerdp_run(freerdp* instance)
        DWORD count;
        HANDLE handles[64];
        DWORD status = WAIT_ABANDONED;
+       STOPWATCH* stopwatch = stopwatch_create();
 
        if (!instance)
                return -1;
@@ -487,8 +489,17 @@ static int wlfreerdp_run(freerdp* instance)
                return -1;
        }
 
+       stopwatch_start(stopwatch);
        while (!freerdp_shall_disconnect(instance))
        {
+               stopwatch_stop(stopwatch);
+               if (stopwatch_get_elapsed_time_in_seconds(stopwatch) > 180)
+               {
+                       //WLog_Print(context->log, WLOG_INFO, "fake mouse move");
+                       freerdp_input_send_mouse_event(instance->input, PTR_FLAGS_MOVE, 800, 600);
+                       stopwatch_reset(stopwatch);
+                       stopwatch_start(stopwatch);
+               }
                handles[0] = context->displayHandle;
                count = freerdp_get_event_handles(instance->context, &handles[1], 63) + 1;
 
```
The above patch is working - even setting the mouse cursor to the same position prevents window from screen locking.

I am also thinking of:
- Detecting somehow session idle state and sending those events only in this case
- Sending the mouse event to the "saved" current cursor position (to not disturb the session in any visible way)
- Add a command line switch for this feature (with parameterized interval)

Now my main question:
If I prepare a PR for this, is it OK for you to merge this kind of feature?

**Comment by akallabeth** [comment]
@manio sounds like a sensible approach (only enabling this if a command line switch is in use, off by default)

from my POV:
1. you can save the cursor position with the last `input_send_mouse_event` (make it part of `rdpInput`?)
2. you can detect idle by remembering the last time some keyboard/mouse event was sent
   * you need a interval to trigger the main loop though to check that periodically
   * there is still the issue of not being able to detect e.g. a video being watched (then mouse movement might be annoying)
1. best implement that in `client/common` library so all clients automatically have this feature
1. best target our current development branch, 2.x series will most likely not get that one as we prepare 3.0 release

**Comment by manio** [comment]
@akallabeth 
Thank you very much!
I'll take all this into account when preparing a PR.
For now I am closing the discussion as we surely review details in the PR itself.
